README.md
## Links

- [简体中文介绍](README.zh-CN.md)

## Introduction

Enjoy Mock is a persistent service that generates mock data quickly and provids
visualization view.

## Features

- Support API proxying
- Convenient shortcuts
- Support Collaborative editing
- Support team project
- Support RESTful
- Support [Swagger](https://swagger.io) | OpenAPI Specification ([1.2](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/1.2.md) & [2.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) & [3.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md))
  - Create project quickly based on Swagger
  - Support displaying parameters and the return value
  - Support displaying class model
- More flexible and extensible in response data
- Support for custom response configuration (example: status/headers/cookies)
- Use [Mock.js](http://mockjs.com/) schema
- Support [restc](https://github.com/ElemeFE/restc) to preview API

## Quick Start

> Before starting, we assume that you're already have installed
> [Node.js](https://nodejs.org) (**v8.x, ~~v10.x is not supported~~**) & [MongoDB](https://www.mongodb.com) (**>= v3.4**) & [Redis](https://redis.io)（**>= v4.0**）.

### Installation

```shell
$ git clone git@github.com:bruceliu68/Enjoy-mock.git
$ cd easy-mock && npm install
```

### Configuration

Find **config/default.json** or create **config/local.json** to overwrite some
configuration.

> Enjoy Mock will load different configuration files according to your
> environment. Reference to [node-config](https://github.com/lorenwest/node-config)
> to get more information because Enjoy Mock uses node-config as its
> configuration module.

```js
{
  "port": 7300,
  "host": "0.0.0.0",
  "pageSize": 30,
  "proxy": false,
  "db": "mongodb://localhost/enjoy-mock",
  "unsplashClientId": "",
  "redis": {
    "keyPrefix": "[Enjoy Mock]",
    "port": 6379,
    "host": "localhost",
    "password": "",
    "db": 0
  },
  "blackList": {
    "projects": [], // projectId, e.g."5a4495e16ef711102113e500"
    "ips": [] // ip, e.g. "127.0.0.1"
  },
  "rateLimit": { // https://github.com/koajs/ratelimit
    "max": 1000,
    "duration": 1000
  },
  "jwt": {
    "expire": "14 days",
    "secret": "shared-secret"
  },
  "upload": {
    "types": [".jpg", ".jpeg", ".png", ".gif", ".json", ".yml", ".yaml"],
    "size": 5242880,
    "dir": "../public/upload",
    "expire": {
      "types": [".json", ".yml", ".yaml"],
      "day": -1
    }
  },
  "ldap": {
    "server": "", // Set server to enable LDAP login. e.g. "ldap://localhost:389" or "ldaps://localhost:389"（use SSL）
    "bindDN": "", // Username，e.g. "cn=admin,dc=example,dc=com"
    "password": "",
    "filter": {
      "base": "", // Base where we can search for users，e.g. "dc=example,dc=com"
      "attributeName": "" // e.g. "mail" or "email" etc.
    }
  },
  "fe": {
    "copyright": "",
    "storageNamespace": "enjoy-mock_",
    "timeout": 25000,
    "publicPath": "/dist/"
  }
}
```

**Note**:

- The default value of `publicPath` is `'/dist/'`. You can replace it to your
  own CDN if necessary.
- If you changed some configuration of `fe`, you should run `build` command
  to adapt that changes.

**Background**:

Enjoy Mock supports two background service,
[Unsplash](https://unsplash.com/developers) and [Bing](http://bing.com).

If you leave `unsplashClientId` blank, the background will be provided by Bing.

### Launch

```sh
$ npm run dev
# Visit http://127.0.0.1:7300
```

## More Commands

```sh
# Build front-end assets
$ npm run build

# Run Enjoy Mock as production environment (You should run `build` first)
$ npm run start

# Run unit test
$ npm run test

# Test lint
$ npm run lint
```

## Deployment

> Please configure your configuration files before this step.

### PM2

We're recommending you to use [PM2](https://github.com/Unitech/pm2) as your
daemon process.

#### Install PM2 Globally

```sh
$ [sudo] npm install pm2 -g
```

#### Launch via PM2

> You should run `build` before this step.

```sh
$ NODE_ENV=production pm2 start app.js
```


CHANGELOG.md
## 更新日志

### 1.6.0
*2018-05-24*

**Bug Fixes**
- 改善 Swagger 同步失败时的提示信息，#144

**New Features**
- 接口数据编辑器支持快捷键保存（win: 'Ctrl-S', mac: 'Command-S'），#139
- redis 支持更多配置项，#157
- 支持 LDAP 登录，#138

### 1.5.1
*2018-01-17*

**Bug Fixes**
- 修复因 `BaseURL` 引起的 404

### 1.5.0
*2018-01-16*
> 从这个版本开始启用 **redis** 并升级至 **koa2**，升级前请注意以下事项。

- redis `(>= v4.0)`
- Node.js `(>= v8.9.1)`
- 配置文件优化后（以 `config/default.json` 为准），移除多余配置（如：APIPrefix）并新增一些配置（如：rateLimit），按需完善即可。

**New Features**
- 支持配置 `proxy`
- 支持配置 `host`
- 支持请求速率配置
- 支持黑名单配置
- 支持上传 `yaml` 文件

**Refactorings**
- 升级至 koa2
- 当团队可用时，新建项目不再默认选中归属

**Bug Fixes**
- 修复在 `proxy` 模式下，`{xx}` 无法被解析的问题
- 修复在接口列表页下无法添加至工作台的问题

### 1.4.0
*2017-12-01*

**New Features**
- 新增上传接口，#52
- 支持配置上传文件的过期时间(天)
- 支持上传 OAS 文档 (JSON/YML)，#70 #112

### 1.3.6
*2017-11-24*

**New Features**
- 支持 OAS 3.0，#6 #120

**Bug Fixes**
- 修复因 `patch` 方式请求导致失败的问题，#121

### 1.3.5
*2017-11-17*

**New Features**
- login 页支持显示版权信息

### 1.3.4
*2017-11-17*

**New Features**
- 支持配置并显示版权信息
- 支持 `Access-Control-Allow-Credentials` 配置

### 1.3.3
*2017-11-06*

**Bug Fixes**
- 修复因 `while (true) {}` 引起的崩溃问题，#113

### 1.3.2
*2017-11-02*

> 注意：该版本更新后，之前的请求参数，响应参数等信息会异常显示（再次同步 Swagger 即可解决）。

**New Features**
- 版本更新时，文档与更新日志会添加徽标提醒

**Bug Fixes**
- 修复由快捷键引起的无法复制的问题，#32

**Refactorings**
- Swagger 相关代码已迁移至 [Swagger Parser Mock](https://github.com/easy-mock/swagger-parser-mock) 模块，#48 #71 #109

### 1.3.1
*2017-10-17*

**Bug Fixes**
- 修复全屏编辑导航栏显示的问题，#98

### 1.3.0
*2017-09-28*

**New Features**
- 支持国际化
- 支持自定义响应配置，#78
- 接口列表页新增返回顶部按钮
- 更新日志新增问题锚点

**Bug Fixes**
- 修复 state 单例的问题
- 修复 jwt.secret 配置错误的问题
- 修复由 bcrypt 引起的无法安装的问题
- 修复团队项目下邀请成员导致 500 的问题，#33
- 修复邀请成员输入框无法匹配用户名的问题，#45
- 修复由 wallpaper 接口引起的 login 页错误的问题，#39
- 修复由系统路径造成的通过 swagger 创建接口引发路径错误的问题

### 1.2.3
*2017-08-31*

**Bug Fixes**
- 修复接口数据正则表达式失效的问题

**New Features**
- 前端新增 JSON 验证

### 1.2.2
*2017-08-30*

**Refactorings**
- 调整 /dashboard 页最近新增用户条数
- 调整 editor 字号为 15px
- 调整接口列表字号

**Bug Fixes**
- 修复无法统计今日接口累积调用的问题

### 1.2.1
*2017-08-29*

**Performance Improvements**

- 优化 unsplash 图片宽度

### 1.2.0
*2017-08-28*

**New Features**
- 支持 restc 方式的预览
- 新增 `/wallpaper` 接口
- 代理接口支持 Restful 风格的参数
- 支持按创建者获取项目列表

**Bug Fixes**
- 修复项目成员无法编辑项目和接口的问题
- 修复当项目和接口 url 为 `/` 时返回 404 的问题
- 修复因数据模型死循环造成内存泄漏的问题

**Performance Improvements**
- 优化统计接口，增加缓存策略

**Refactorings**
- 前端整体重构
- 接口 `/proxy/cross` 调整为 `/proxy`

### 1.1.11

- 增加 `easy-mock-cli` 导航
- 支持向数据模型的 Function 中注入 Mock 对象
- 重构接口无某类型数据时隐藏相应 Tab
- 优化文档

### 1.1.10

**Bug Fixes**
- 由于缺少 `definitions` 字段导致无法创建接口的问题
- 无法通过 `/project/create/:groupId/by_swagger` 创建项目的问题

### 1.1.9

**Bug Fixes**
- 基于 Swagger 创建 Mock 造成栈溢出的问题
- 参数 Model 被重复创建的问题

### 1.1.8

**New Features**
- 全屏模式编辑下支持 `保存` `格式化` 操作
- 静态资源走 CDN

**Bug Fixes**
- 基于 Swagger 创建 Mock 造成栈溢出的问题
- Swagger allOf 属性无法生成完整数据结构的问题

**Refactorings**
- 基于 Swagger 创建 Mock 的实现
- 基于 Swagger 创建 Mock 失败时将返回具体错误信息
- 团队搜索采用绝对匹配

### 1.1.7

**New Features**
- 添加 _req.params 对象

### 1.1.6

**Bug Fixes**
- 无法基于 Swagger 2.0 创建的问题
- 当 Mock Url 为 / 时无法访问的问题

### 1.1.5

**New Features**
- 查看数据使用量
- 在 Swagger Docs API 旁添加文档锚点

**Bug Fixes**
- BaseUrl 不支持2级以上的问题
- 数据模型格式化失败的问题
- 无法基于 Swagger 2.0 创建的问题

**Refactorings**
- 点击按钮退出全屏

### 1.1.4

**New Features**
- 登录页可取消自动创建用户的流程
- 从统计中跳转进项目

**Bug Fixes**
- 不支持花括号分隔符的问题
- 因包版本不同导致无法编译的问题
- 搜索无法统计 Mock 使用量的问题

### 1.1.3

**Bug Fixes**
- 头像上传没有去掉 `URL` 前缀的问题

### 1.1.2

**New Features**
- 支持根据 `响应数据` 生成 `JavaScript Class Model`
- 支持查看 `Mock 排行榜`

### 1.1.1

**Refactorings**
- `/project/create/:groupId/by_swagger` 支持更新操作

### 1.1.0

**New Features**
- 支持 `JSONP`

**Bug Fixes**
- `swagger_url` 不支持下划线的问题

### 1.0.0

**Bug Fixes**
- 无法更新头像的问题
- `Element UI` 造成的服务端渲异常问题

**Performance Improvements**
- Client
    - 优化体积大小

### 1.0.0-rc.9

**New Features**
- 支持在 `Mock Handler Function` 里获取 `Request` 对象

### 1.0.0-rc.8

**Bug Fixes**
- 无法更新接口参数的问题
- 无法支持 `Swagger 1.2 Model` 为数组的问题

### 1.0.0-rc.7

**New Features**
- 支持基于 `Swagger 2.0` 的创建
- 支持基于 `Swagger` 创建的参数显示
- 支持全屏编辑数据
- 支持数据预加载

**Bug Fixes**
- 无法上传头像的问题

**Refactorings**
- 重构用户身份信息过期的授权方式

### 1.0.0-rc.6

**Bug Fixes**
- 当 `Swagger` 某接口字段类型异常时无法生成文档的问题

**Performance Improvements**
- Server
    - 基于 `Swagger` 创建项目方式的实现
- Client
    - 基于 `Swagger` 创建项目时，不再支持传递 `Cookie`

### 1.0.0-rc.5

**New Features**
- 支持`团队管理`
- 支持`团队类型项目`
- 支持复制团队项目到个人项目
- 以 `Swagger` 创建团队项目

**Bug Fixes**
- 以 `/api` 开头出现的授权失效问题

**Performance Improvements**
- Client
    - 预览
    - 面包屑
    - 文档

### 1.0.0-rc.4

**New Features**
- 新增通过 `Project Ids` 获取 Mock 列表

### 1.0.0-rc.3

**New Features**
- 新增工作台
- 新增文档
- 新增 `Loading` 状态
- 支持 `HTTPS`

**Bug Fixes**
- 修复因窗口宽度导致元素展示不正确的问题
- 修复 Mock 不支持正则和方法的问题
- 修复更新接口查重不正确的问题

**Refactorings**
- 重构 Swagger 更新逻辑，支持增量覆盖式更新
- Project Mock 支持批量插入
- 替换字段 `id` 为 `_id`
- 去掉外键的 `_id`

**Performance Improvements**
- Server
    - 优化删除接口，支持批量删除
    - 添加索引
- Client
    - 优化布局
    - `项目` 和 `Mock` 更新后同步本地数据

### 1.0.0-rc.2

**New Features**
- 增加 `X-Request-Id` 响应头
- 接入单点登录，内网允许并且只允许通过单点登录完成授权
- 增加用户注销接口
- 增加分页显示
- 增加数据统计
- 增加数据导出

**Bug Fixes**
- 修复由 ElementUI 引起的 SSR 渲染失败的问题
- 修复 Safari 上的一些 Bug

**Refactorings**
- Mock Router
- 身份信息持久化

**Performance Improvements**
- Server
    - 优化异常处理
    - 优化响应数据结构
    - 入参验证
    - 错误提示
    - `token` 过期时间变为 `7天`
    - 优化头像生成策略
- Client
    - 配合响应数据结构的优化
    - 介绍页头图切换源地址，加快响应速度
    - 取消编辑后自动关闭弹窗的特性
    - 优化部分页面的元素位置与操作逻辑
    - 优化搜索体验

### 1.0.0-rc.1

**New Features**
- 支持快速复制以创建新项目
- 新增主页，用于介绍 Enjoy Mock

**Performance Improvements**
- Server
    - 新增代理接口，可实现 `get` 方式的跨域请求
    - `token` 过期时间变为 `2小时`
- Client
    - 完善面包屑导航
    - 优化登录背景获取方式
    - 优化Mock页的创建名称
    - 设置请求超时为 `25s`
    - 优化编辑页头像显示方式
    - 替换头像上传接口

### 1.0.0-rc.0

**New Features**
- 支持创建以项目为单位的 `Mock List`
- 支持简单的数据模型( 纯 `JSON` 对象)
- 支持以 [Mock.js](http://mockjs.com/) 为基础的数据模型
- 支持结合 `Swagger` 快速创建 `Mock List`
- 支持接口代理
- 支持快捷键创建 `Project` 和 `Mock`
- 支持邀请成员，协同编辑
- 支持 `Restful`
- 支持快速复制以创建新接口


views/locale/en.js
export default {
  confirm: {
    title: 'Notification'
  },
  p: {
    dashboard: {
      header: {
        title: 'Dashboard',
        description: 'Display your application usage in real time.'
      },
      total: {
        user: 'Total Users | ',
        mock: 'Total Mocks | ',
        project: 'Total Projects | ',
        mockUse: 'Total Mock Usage | '
      },
      today: {
        user: 'New Users | ',
        mock: 'New Mocks | ',
        project: 'New Projects | ',
        mockUse: 'Mock Usage Today | '
      }
    },
    docs: {
      header: {
        title: 'Docs | Release Notes',
        description: 'Read docs and use Enjoy Mock like pro. | Never stop updating.'
      }
    },
    group: {
      header: {
        title: 'Team Projects',
        description: 'Teams you created or joined.'
      },
      modal: {
        tab: {
          create: 'Create New Team | Name | Make your team name spical to avoid accidently access',
          join: 'Join a Team | Name | Please type team name',
          edit: 'Edit Team | Name | Make your team name spical to avoid accidently access'
        }
      },
      placeholder: 'Not matched team founded. | Create your first team to start cooperating :)',
      keyboards: [{
        category: 'Action',
        list: 'Create New Teams'
      }],
      confirm: {
        delete: {
          content: 'You will <strong>dismiss</strong> or <strong>quit</strong> team <strong>{name}</strong> , are you sure?'
        }
      },
      create: {
        success: 'Create success'
      },
      update: {
        success: 'Update success'
      },
      remove: {
        success: 'Remove success'
      },
      join: {
        success: 'Alread joined {groupName}',
        warning: '{groupName} not exsit'
      }
    },
    login: {
      description: 'Mock more efficiently | ',
      form: {
        button: 'Let\'go | Login',
        placeholder: 'Username, will register if user doesn\'t exsit | LDAP Username',
        password: 'Password'
      },
      about: 'What is Enjoy Mock? | A vitrual API mock up tool, and generate mock up data.',
      feature: [
        'Swagger | Kickass feature, using Swagger to create Mock API in few seconds, increase efficiency up to 99%.',
        'Mock.js | Maybe you used this tool, greate! Now Enjoy Mock integrated Mock.js, it mockes up data more fast.',
        'Docs | We provide comprehensive documents to help you master Enjoy Mock. Contact us, if you found bugs or have feedback.',
        'Responsive Data | Enjoy Mock suport responsive data. It reponses data based on request query parameters.',
        'Syntax Lint | Forget syntax of Mock.js? No problem. Now type em in editor for syntax lint.'
      ],
      confirm: {
        register: {
          content: 'User does\'t exsit, do you want create a new user with this username and password? Notice: You cannot retrieve your password if you forget in this version',
          success: 'Register Success'
        }
      }
    },
    new: {
      header: {
        title: 'Create New Project',
        description: 'Start your next amazing new project'
      },
      form: {
        name: 'Belongs To / Project Name | Make it shot and expressive, e.g.：petstore',
        url: 'Base URL | Make it shot and expressive, e.g.：/nba',
        description: 'Description | Will be project name if null',
        swagger: 'Swagger Docs API | Optional | If backend provides Swagger Doc (without authentication error), Enjoy Mock will create Mock based on Swagger API address you provide.',
        member: ['Ask people join for collaboration', 'Optional | Lose effectiveness under team project', 'Support fuzzy search, user nickname, username'],
        confirm: 'Please type in the name of the project to confirm. | Project name | This action <strong>CANNOT<strong> be undone. This will permanently delete project:',
        button: {
          delete: 'Delete Project',
          create: 'Create Project',
          update: 'Save'
        },
        success: {
          create: 'Create Success',
          update: 'Update Success',
          delete: '{name} is deleted'
        },
        error: {
          groupIsNull: 'Please select the project belongs to'
        }
      },
      formatError: 'File only supports JSON/YML format',
      uploadSuccess: 'Upload Success | Upload Success, The file will expire at {date}'
    },
    profile: {
      header: {
        title: 'Edit Profile',
        description: 'Personalize your profile'
      },
      modal: {
        title: 'View Profile Picture'
      },
      form: {
        language: 'Language',
        nickName: 'Nick Name',
        password: 'Password',
        passwordCheck: 'Confirm password',
        update: 'Update Profile',
        avatar: 'Profile Picture',
        upload: 'Upload New Picture'
      },
      formatError: 'Unsupported picture formate | {name} is not support, please upload *.jpg or *.png picture.',
      updateSuccess: 'Update Success | Please login again to check changes.',
      validateError: 'Password is not matching!'
    },
    project: {
      header: {
        title: ['Personal Project', '{groupName}', 'Work Station'],
        description: ['Here are all your personal projects, also projects you are working with.', 'Welcome join group {groupName}, Let\'s Mock up together.', 'Add projects to Work Station for faster access.']
      },
      placeholder: ['Wanna finish work faster? Let\'s Mock up。', 'Team Project is perfect for cooperation.', 'Wanna work faster? Add to projects to Work Station.', 'No match founded'],
      filter: ['All', 'Created', 'Joined'],
      cloneSuccess: 'Copy Success',
      copySuccess: 'Project address is in clipboard',
      deleteSuccess: '{name} is deleted',
      modal: {
        delete: {
          title: 'Confirm Delete',
          description: 'Please type in the name of the project to confirm. | This action <strong>CANNOT<strong> be undone. This will permanently delete project:',
          button: 'Delete',
          placeholder: 'Project name'
        }
      },
      control: ['Copy Project Address', 'Copy Project', 'Delete Project']
    },
    detail: {
      header: {
        description: ['Personal Project', 'Team Project']
      },
      nav: ['Mock List', 'Setting'],
      workbench: 'Work Station',
      download: 'Downlaod as zip file | Download',
      member: 'Member List',
      keyboards: [
        { category: 'Navigation' },
        { category: 'Action', list: ['Create new Mock', 'Add / Remove Work Station', 'Sync with Swagger'] }
      ],
      columns: ['Description', 'Action'],
      action: ['Preview Mock', 'Edit Mock', 'Copy Mock Address', 'Copy', 'Delete'],
      copySuccess: 'Project address is in clipboard',
      syncSwagger: {
        action: 'Sync Swagger',
        warning: 'Please set up Swagger API address in settings page',
        confirm: 'This will sync Swagger API, do you want contiune?',
        success: 'Sync Success',
        syncResult: '同步结果',
        syncFailed: {
          title: '部分接口同步失败',
          desc: '接口中存在语法错误，请检查是否为标准 JSON 格式（例：被忽略的双引号、定义方法等）。'
        }
      },
      remove: {
        confirm: ['This action <strong>CANNOT<strong> be undone. Are you sure want to delete selected row?', 'This action <strong>CANNOT<strong> be undone. Do you want contiune?'],
        success: 'Delete Success'
      },
      create: {
        action: 'Create Mock',
        success: 'Create Success'
      },
      expand: {
        description: 'Description',
        tab: ['Request Parameters', 'Response Parameters'],
        columnsRequest: ['Parameter', 'Description', 'Parameter Type', 'Data Type'],
        columnsResponse: ['Status Code', 'Description'],
        defaultDescription: 'Too layz to inlude a description'
      },
      editor: {
        title: ['Update Mock', 'Create New Mock'],
        action: ['Updae', 'Create'],
        autoClose: 'Auto Close',
        control: ['Format', 'Preview', 'Close'],
        submit: {
          error: ['Mock data can\'t be null', 'Please check data definiation is correct.'],
          updateSuccess: 'Update Success'
        }
      }
    }
  },
  c: {
    keyboardShort: {
      modalTitle: 'Keyboard Shortcuts',
      keyboards: [
        { category: 'Global Navigations', list: ['Personal Projects', 'Team Projects', 'Work Station', 'Docs'] },
        { category: 'Global Actions', list: ['Create New Project', 'Search'] }
      ]
    },
    layout: {
      menu: [
        ['Project', 'Personal Project', 'Team Project'],
        'Work Station', 'Dashboard', 'Docs',
        ['More', 'API Generator', 'Mock Syntax'],
        ['Edit Profile', 'Logout', 'Login']
      ]
    },
    logOut: {
      text: 'Redirecting...'
    }
  }
}


controllers/mock.js
'use strict'

const _ = require('lodash')
const { VM } = require('vm2')
const nodeURL = require('url')
const JSZip = require('jszip')
const Mock = require('mockjs')
const axios = require('axios')
const config = require('config')
const pathToRegexp = require('path-to-regexp')

const util = require('../util')
const ft = require('../models/fields_table')
const { MockProxy, ProjectProxy, UserGroupProxy } = require('../proxy')

const redis = util.getRedis()
const defPageSize = config.get('pageSize')

async function checkByMockId (mockId, uid) {
  const api = await MockProxy.getById(mockId)

  if (!api) return '接口不存在'

  const project = await checkByProjectId(api.project.id, uid)

  if (typeof project === 'string') return project
  return { api, project }
}

async function checkByProjectId (projectId, uid) {
  const project = await ProjectProxy.findOne({ _id: projectId })

  if (project) {
    const group = project.group
    if (group) {
      const userGroup = await UserGroupProxy.findOne({ user: uid, group: group })
      if (!userGroup) return '无权限操作'
    } else if (project.user.id !== uid) {
      /* istanbul ignore else */
      if (!_.find(project.members, ['id', uid])) return '无权限操作'
    }
    return project
  }

  return '项目不存在'
}

module.exports = class MockController {
  /**
   * 创建接口
   * @param Object ctx
   */

  static async create (ctx) {
    const uid = ctx.state.user.id
    const mode = ctx.checkBody('mode').notEmpty().value
    const projectId = ctx.checkBody('project_id').notEmpty().value
    const description = ctx.checkBody('description').notEmpty().value
    const url = ctx.checkBody('url').notEmpty().match(/^\/.*$/i, 'URL 必须以 / 开头').value
    const method = ctx.checkBody('method').notEmpty().toLow().in(['get', 'post', 'put', 'delete', 'patch']).value

    if (ctx.errors) {
      ctx.body = ctx.util.refail(null, 10001, ctx.errors)
      return
    }

    const project = await checkByProjectId(projectId, uid)

    if (typeof project === 'string') {
      ctx.body = ctx.util.refail(project)
      return
    }

    const api = await MockProxy.findOne({
      project: projectId,
      url,
      method
    })

    if (api) {
      ctx.body = ctx.util.refail('请检查接口是否已经存在')
      return
    }

    await MockProxy.newAndSave({
      project: projectId,
      description,
      method,
      url,
      mode
    })

    await redis.del('project:' + projectId)
    ctx.body = ctx.util.resuccess()
  }

  /**
   * 获取接口列表
   * @param Object ctx
   */

  static async list (ctx) {
    const uid = ctx.state.user.id
    const keywords = ctx.query.keywords
    const projectId = ctx.checkQuery('project_id').notEmpty().value
    const pageSize = ctx.checkQuery('page_size').empty().toInt().gt(0).default(defPageSize).value
    const pageIndex = ctx.checkQuery('page_index').empty().toInt().gt(0).default(1).value

    if (ctx.errors) {
      ctx.body = ctx.util.refail(null, 10001, ctx.errors)
      return
    }

    const opt = {
      skip: (pageIndex - 1) * pageSize,
      limit: pageSize,
      sort: '-create_at'
    }

    const where = { project: projectId }

    if (keywords) {
      const keyExp = new RegExp(keywords)
      where.$or = [{
        url: keyExp
      }, {
        description: keyExp
      }, {
        method: keyExp
      }, {
        mode: keyExp
      }]
    }

    let mocks = await MockProxy.find(where, opt)
    let project = await ProjectProxy.getById(uid, projectId)

    /* istanbul ignore else */
    if (project) {
      project.members = project.members.map(o => _.pick(o, ft.user))
      project.extend = _.pick(project.extend, ft.projectExtend)
      project.group = _.pick(project.group, ft.group)
      project.user = _.pick(project.user, ft.user)
      project = _.pick(project, ['user'].concat(ft.project))
    }

    mocks = mocks.map(o => _.pick(o, ft.mock))

    ctx.body = ctx.util.resuccess({ project: project || {}, mocks })
  }

  /**
   * 更新接口
   * @param Object ctx
   */

  static async update (ctx) {
    const uid = ctx.state.user.id
    const id = ctx.checkBody('id').notEmpty().value
    const mode = ctx.checkBody('mode').notEmpty().value
    const description = ctx.checkBody('description').notEmpty().value
    const url = ctx.checkBody('url').notEmpty().match(/^\/.*$/i, 'URL 必须以 / 开头').value
    const method = ctx.checkBody('method').notEmpty().toLow().in(['get', 'post', 'put', 'delete', 'patch']).value

    if (ctx.errors) {
      ctx.body = ctx.util.refail(null, 10001, ctx.errors)
      return
    }

    const result = await checkByMockId(id, uid)

    if (typeof result === 'string') {
      ctx.body = ctx.util.refail(result)
      return
    }

    const { api, project } = result

    api.url = url
    api.mode = mode
    api.method = method
    api.description = description

    const existMock = await MockProxy.findOne({
      _id: { $ne: api.id },
      project: project.id,
      url: api.url,
      method: api.method
    })

    if (existMock) {
      ctx.body = ctx.util.refail('接口已经存在')
      return
    }

    await MockProxy.updateById(api)
    await redis.del('project:' + project.id)
    ctx.body = ctx.util.resuccess()
  }

  /**
   * 获取 Mock 接口
   * @param {*} ctx
   */

  static async getMockAPI (ctx) {
    const { query, body } = ctx.request
    const method = ctx.method.toLowerCase()
    const jsonpCallback = query.jsonp_param_name && (query[query.jsonp_param_name] || 'callback')
    let { projectId, mockURL } = ctx.pathNode
    const redisKey = 'project:' + projectId
    let apiData, apis, api

    apis = await redis.get(redisKey)

    if (apis) {
      apis = JSON.parse(apis)
    } else {
      apis = await MockProxy.find({ project: projectId })
      if (apis[0]) await redis.set(redisKey, JSON.stringify(apis), 'EX', 60 * 30)
    }

    if (apis[0] && apis[0].project.url !== '/') {
      mockURL = mockURL.replace(apis[0].project.url, '') || '/'
    }

    api = apis.filter((item) => {
      const url = item.url.replace(/{/g, ':').replace(/}/g, '') // /api/{user}/{id} => /api/:user/:id
      return item.method === method && pathToRegexp(url).test(mockURL)
    })[0]

    if (!api) ctx.throw(404)

    Mock.Handler.function = function (options) {
      const mockUrl = api.url.replace(/{/g, ':').replace(/}/g, '') // /api/{user}/{id} => /api/:user/:id
      options.Mock = Mock
      options._req = ctx.request
      options._req.params = util.params(mockUrl, mockURL)
      options._req.cookies = ctx.cookies.get.bind(ctx)
      return options.template.call(options.context.currentContext, options)
    }

    if (/^http(s)?/.test(api.mode)) { // 代理模式
      const url = nodeURL.parse(api.mode.replace(/{/g, ':').replace(/}/g, ''), true)
      const params = util.params(api.url.replace(/{/g, ':').replace(/}/g, ''), mockURL)
      const pathname = pathToRegexp.compile(url.pathname)(params)
      try {
        apiData = await axios({
          method: method,
          url: url.protocol + '//' + url.host + pathname,
          params: _.assign({}, url.query, query),
          data: body,
          timeout: 3000,
          headers: ctx.headers
        }).then(res => res.data)
      } catch (error) {
        ctx.body = ctx.util.refail(error.message || '接口请求失败')
        return
      }
    } else {
      const vm = new VM({
        timeout: 1000,
        sandbox: {
          Mock: Mock,
          mode: api.mode,
          template: new Function(`return ${api.mode}`) // eslint-disable-line
        }
      })

      vm.run('Mock.mock(new Function("return " + mode)())') // 数据验证，检测 setTimeout 等方法
      apiData = vm.run('Mock.mock(template())') // 解决正则表达式失效的问题

      /* istanbul ignore else */
      if (apiData._res) { // 自定义响应 Code
        let _res = apiData._res
        ctx.status = _res.status || /* istanbul ignore next */ 200
        /* istanbul ignore else */
        if (_res.cookies) {
          for (let i in _res.cookies) {
            /* istanbul ignore else */
            if (_res.cookies.hasOwnProperty(i)) ctx.cookies.set(i, _res.cookies[i])
          }
        }
        /* istanbul ignore next */
        if (_res.headers) {
          for (let i in _res.headers) {
            /* istanbul ignore next */
            if (_res.headers.hasOwnProperty(i)) ctx.set(i, _res.headers[i])
          }
        }
        /* istanbul ignore next */
        if (_res.status && parseInt(_res.status, 10) !== 200 && _res.data) apiData = _res.data
        delete apiData['_res']
      }
    }

    await redis.lpush('mock.count', api._id)
    if (jsonpCallback) {
      ctx.type = 'text/javascript'
      ctx.body = `${jsonpCallback}(${JSON.stringify(apiData, null, 2)})`
        .replace(/\u2028/g, '\\u2028')
        .replace(/\u2029/g, '\\u2029') // JSON parse vs eval fix. https://github.com/rack/rack-contrib/pull/37
    } else {
      ctx.body = apiData
    }
  }

  /**
   * Enjoy Mock CLI 依赖该接口获取接口数据
   * @param Object ctx
   */

  static async getAPIByProjectIds (ctx) {
    let projectIds = ctx.checkQuery('project_ids').notEmpty().value

    if (ctx.errors) {
      ctx.body = ctx.util.refail(null, 10001, ctx.errors)
      return
    }

    projectIds = projectIds.split(',')

    const apis = await MockProxy.find({
      project: {
        $in: projectIds
      }
    })

    const projects = await ProjectProxy.findByIds(projectIds)

    const result = {}

    projects.forEach((project) => {
      const projectId = project.id
      let newMocks = apis.filter(o => (o.project.id === projectId))
      let newProject = projects.filter(o => (o.id === projectId))[0]

      newProject.members = newProject.members.map(o => _.pick(o, ft.user))
      newProject.user = _.pick(newProject.user, ft.user)
      newProject = _.pick(newProject, ['user'].concat(ft.project))
      newMocks = newMocks.map(o => _.pick(o, ft.mock))

      result[projectId] = {
        project: newProject,
        mocks: newMocks
      }
    })

    ctx.body = ctx.util.resuccess(result)
  }

  /**
   * 接口导出
   * @param Object ctx
   */

  static async exportAPI (ctx) {
    const zip = new JSZip()
    const ids = ctx.checkBody('ids').empty().type('array').value
    const projectId = ctx.checkBody('project_id').empty().value
    let apis

    if (ctx.errors) {
      ctx.body = ctx.util.refail(null, 10001, ctx.errors)
      return
    }

    if (projectId) {
      apis = await MockProxy.find({ project: projectId })
    } else if (!_.isEmpty(ids)) {
      apis = await MockProxy.find({
        _id: {
          $in: ids
        }
      })
    } else {
      ctx.body = ctx.util.refail('参数不能为空')
      return
    }

    if (_.isEmpty(apis)) {
      ctx.body = ctx.util.refail('没有可导出的接口')
      return
    }

    apis.forEach((api) => {
      zip.file(`${api.project.url}${api.url}.json`, api.mode)
    })

    const content = await zip.generateAsync({ type: 'nodebuffer' })

    ctx.set('Content-disposition', 'attachment; filename=Enjoy-Mock-API.zip')
    ctx.body = content
  }

  /**
   * 删除接口
   * @param Object ctx
   */

  static async delete (ctx) {
    const uid = ctx.state.user.id
    const projectId = ctx.checkBody('project_id').notEmpty().value
    const ids = ctx.checkBody('ids').notEmpty().type('array').value

    if (ctx.errors) {
      ctx.body = ctx.util.refail(null, 10001, ctx.errors)
      return
    }

    const project = await checkByProjectId(projectId, uid)

    if (typeof project === 'string') {
      ctx.body = ctx.util.refail(project)
      return
    }

    await MockProxy.find({
      _id: {
        $in: ids
      },
      project: projectId
    })

    await MockProxy.delByIds(ids)
    await redis.del('project:' + projectId)
    ctx.body = ctx.util.resuccess()
  }
}
